Too long Args lead to compilation failure

[ERROR] Failed to execute goal com.github.wvengen:proguard-maven-plugin:2.2.0:proguard (default) on project xxx-xx-xxxx: Obfuscation failed (result=1) -> [Help1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal com.github.wvengen:proguard-maven-plugin:2.2.0:proguard (default) on project xxx-xx-xxxx: Obfuscation failed (result=1)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: Obfuscation failed (result=1)
        at com.github.wvengen.maven.proguard.ProGuardMojo.proguardMain(ProGuardMojo.java:909)
        at com.github.wvengen.maven.proguard.ProGuardMojo.execute(ProGuardMojo.java:696)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
        ... 20 more
[ERROR]
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command

[ERROR]   mvn <goals> -rf :xxx-xx-xxxx

Configuration:
```
<plugin>
	<groupId>com.github.wvengen</groupId>
	<artifactId>proguard-maven-plugin</artifactId>
	<version>2.2.0</version>
	<executions>
		<execution>
			<phase>package</phase>
			<goals>
				<goal>proguard</goal>
			</goals>
		</execution>
	</executions>
	<configuration>
		<obfuscate>true</obfuscate>
		<proguardInclude>src/main/assembly/proguard.conf</proguardInclude>
	        <libs>                       
			<lib>${java.home}/jmods/java.base.jmod</lib>
			<lib>${java.home}/jmods/java.datatransfer.jmod</lib>
			<lib>${java.home}/jmods/java.desktop.jmod</lib>
			<lib>${java.home}/jmods/java.instrument.jmod</lib>
			<lib>${java.home}/jmods/java.logging.jmod</lib>
			<lib>${java.home}/jmods/java.management.jmod</lib>
			<lib>${java.home}/jmods/java.management.rmi.jmod</lib>
			<lib>${java.home}/jmods/java.naming.jmod</lib>
			<lib>${java.home}/jmods/java.net.http.jmod</lib>
			<lib>${java.home}/jmods/java.prefs.jmod</lib>
			<lib>${java.home}/jmods/java.rmi.jmod</lib>
			<lib>${java.home}/jmods/java.scripting.jmod</lib>
			<lib>${java.home}/jmods/java.se.jmod</lib>
			<lib>${java.home}/jmods/java.security.jgss.jmod</lib>
			<lib>${java.home}/jmods/java.compiler.jmod</lib>
			<lib>${java.home}/jmods/java.security.sasl.jmod</lib>
			<lib>${java.home}/jmods/java.smartcardio.jmod</lib>
			<lib>${java.home}/jmods/java.sql.jmod</lib>
			<lib>${java.home}/jmods/java.sql.rowset.jmod</lib>
			<lib>${java.home}/jmods/java.transaction.xa.jmod</lib>
			<lib>${java.home}/jmods/java.xml.crypto.jmod</lib>
			<lib>${java.home}/jmods/java.xml.jmod</lib>
			<lib>${java.home}/jmods/jdk.accessibility.jmod</lib>
			<lib>${java.home}/jmods/jdk.aot.jmod</lib>
			<lib>${java.home}/jmods/jdk.attach.jmod</lib>
			<lib>${java.home}/jmods/jdk.charsets.jmod</lib>
			<lib>${java.home}/jmods/jdk.compiler.jmod</lib>
			<lib>${java.home}/jmods/jdk.dynalink.jmod</lib>
			<lib>${java.home}/jmods/jdk.httpserver.jmod</lib>
			<lib>${java.home}/jmods/jdk.internal.ed.jmod</lib>
			<lib>${java.home}/jmods/jdk.internal.jvmstat.jmod</lib>
			<lib>${java.home}/jmods/jdk.internal.le.jmod</lib>
			<lib>${java.home}/jmods/jdk.internal.opt.jmod</lib>
			<lib>${java.home}/jmods/jdk.jartool.jmod</lib>
			<lib>${java.home}/jmods/jdk.javadoc.jmod</lib>
			<lib>${java.home}/jmods/jdk.jconsole.jmod</lib>
			<lib>${java.home}/jmods/jdk.jdi.jmod</lib>
			<lib>${java.home}/jmods/jdk.jdwp.agent.jmod</lib>
			<lib>${java.home}/jmods/jdk.jfr.jmod</lib>
			<lib>${java.home}/jmods/jdk.jshell.jmod</lib>
			<lib>${java.home}/jmods/jdk.jsobject.jmod</lib>
			<lib>${java.home}/jmods/jdk.management.agent.jmod</lib>
			<lib>${java.home}/jmods/jdk.net.jmod</lib>
			<lib>${java.home}/jmods/jdk.scripting.nashorn.jmod</lib>
			<lib>${java.home}/jmods/jdk.sctp.jmod</lib>
			<lib>${java.home}/jmods/jdk.security.auth.jmod</lib>
			<lib>${java.home}/jmods/jdk.security.jgss.jmod</lib>
			<lib>${java.home}/jmods/jdk.unsupported.desktop.jmod</lib>
			<lib>${java.home}/jmods/jdk.unsupported.jmod</lib>
			<lib>${java.home}/jmods/jdk.xml.dom.jmod</lib>
		</libs>
		<addMavenDescriptor>false</addMavenDescriptor>                  
	</configuration>
	<dependencies> 
	   <dependency>
			<groupId>net.sf.proguard</groupId>
			<artifactId>proguard-base</artifactId>
			<version>6.2.0</version>
	   </dependency>
	</dependencies>
</plugin> 
 ```

--- a/src/main/java/com/github/wvengen/maven/proguard/ProGuardMojo.java
+++ b/src/main/java/com/github/wvengen/maven/proguard/ProGuardMojo.java
@@ -134,7 +134,10 @@ public class ProGuardMojo extends AbstractMojo {
 	private boolean putLibraryJarsInTempDir;
 
 	/**
-	 * Create a temporary configuration file
+	 * Use this parameter if your command line arguments become too long and execution fails.
+     *
+     * If this parameter is 'true', the configuration is passed to the proguard process through a file, instead of through
+     * command line arguments. This bypasses the operating system restrictions on the length of the command line arguments.
 	 *
 	 * @parameter default-value="false"
 	 */

--- a/src/main/java/com/github/wvengen/maven/proguard/ProGuardMojo.java
+++ b/src/main/java/com/github/wvengen/maven/proguard/ProGuardMojo.java
@@ -21,6 +21,7 @@
 package com.github.wvengen.maven.proguard;
 
 import java.io.File;
+import java.io.FileWriter;
 import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.IOException;
@@ -131,6 +132,19 @@ public class ProGuardMojo extends AbstractMojo {
 	 * @parameter default-value="false"
 	 */
 	private boolean putLibraryJarsInTempDir;
+
+	/**
+	 * Create a temporary configuration file
+	 *
+	 * @parameter default-value="false"
+	 */
+	private boolean generateTemporaryConfigurationFile;
+
+	/**
+	 * @parameter expression="${project.build.directory}/generated-proguard.conf"
+	 * @readonly
+	 */
+	private File temporaryConfigurationFile;
 
 	/**
 	 * Specifies that project compile dependencies should be added as injar.
@@ -691,6 +705,32 @@ public class ProGuardMojo extends AbstractMojo {
 
 		if (options != null) {
 			Collections.addAll(args, options);
+		}
+
+
+		if(generateTemporaryConfigurationFile) {
+			log.info("building config file");
+
+			StringBuilder stringBuilder = new StringBuilder();
+			for (String arg : args) {
+				if (arg.startsWith("-")) {
+					stringBuilder.append("\n");
+				}else{
+					stringBuilder.append(" ");
+				}
+				stringBuilder.append(arg);
+			}
+
+			try {
+				FileWriter writer = new FileWriter(temporaryConfigurationFile);
+				IOUtils.write(stringBuilder.toString(), writer);
+				writer.close();
+			} catch (IOException e) {
+				throw new MojoFailureException("cannot write to temporary configuration file " + temporaryConfigurationFile, e);
+			}
+			args = new ArrayList<String>();
+			args.add("-include");
+			args.add(fileToString(temporaryConfigurationFile));
 		}
 
 		log.info("execute ProGuard " + args.toString());

--- a/src/main/java/com/github/wvengen/maven/proguard/ProGuardMojo.java
+++ b/src/main/java/com/github/wvengen/maven/proguard/ProGuardMojo.java
@@ -715,19 +715,24 @@ public class ProGuardMojo extends AbstractMojo {
 			for (String arg : args) {
 				if (arg.startsWith("-")) {
 					stringBuilder.append("\n");
-				}else{
+				} else {
 					stringBuilder.append(" ");
 				}
 				stringBuilder.append(arg);
 			}
 
+			FileWriter writer = null;
 			try {
-				FileWriter writer = new FileWriter(temporaryConfigurationFile);
+				writer = new FileWriter(temporaryConfigurationFile);
 				IOUtils.write(stringBuilder.toString(), writer);
-				writer.close();
 			} catch (IOException e) {
 				throw new MojoFailureException("cannot write to temporary configuration file " + temporaryConfigurationFile, e);
-			}
+			} finally {
+			    if(writer != null) {
+                    writer.close();
+                }
+            }
+
 			args = new ArrayList<String>();
 			args.add("-include");
 			args.add(fileToString(temporaryConfigurationFile));

--- a/src/site/apt/index.apt
+++ b/src/site/apt/index.apt
@@ -124,7 +124,7 @@ proguard-maven-plugin
 
         If you have a huge list of dependencies the list of <<<-libraryjars>>> the resulting command line to execute ProGaurd could become too long. On Windows the error message could look like <<<CreateProcess error=206, The filename or extension is too long>>>.
 
-        * <<< <putLibraryJarsInTempDir>true</putLibraryJarsInTempDir> >>> makes the plugin copy all the library jars to a single temporary directory and pass that directory as the only <<<-libraryjars>>> argument to ProGuard. Build performance will be a bit worse, but the command line will be much shorter.
+        * <<< <generateTemporaryConfigurationFile>true</generateTemporaryConfigurationFile> >>> makes the plugin pass the configuration by a temporary file instead of over the command line. Build performance should not be impacted by this.
 
 
 * Usage

